Added general infrastructure for handling target typing for expressions in C# binder. Part 1.

AlekseyTs opened the pull request:
- This change puts the required APIs in place.
- Changes tuples implementation to take advantage of the infrastructure.
- Makes appropriate changes to the binding pipeline just enough to handle scenarios
covered by existing tuples tests.

@VSadov, @gafter, @jcouv, @dotnet/roslyn-compiler Please review.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder.cs:
```diff
@@ -682,6 +682,17 @@ private static ThreeState ReportDiagnosticsIfObsoleteInternal(DiagnosticBag diag
         internal enum BindValueKind : byte
         {
             /// <summary>
+            /// Same as <see cref="RValue"/>, but in context where there is a target type for it.
```
Please write up a summary describing the kind of specification change you are contemplating (i.e. the "infrastructure" that would be required in the C# language specification, and how and where it would be used). We should have LDM approval before integrating this (as opposed to the "traditional" approach of allowing a tuple expression to be of its natural type).

AlekseyTs commented:
I am going to add commits to this PR as I am making more progress.

gafter commented:
Sounds good. I've reviewed the code, and don't have any issues with it, other than the fact that it is implementing something that hasn't been contemplated as a spec change.

AlekseyTs commented:
I've pushed a new commit to this PR:
Added general infrastructure for handling target typing for expressions in C# binder. Part 2.

- Addressed failures of tuple related refactoring tests. This introduced a new API, 
SemanticModel.GetNaturalType so that consumers can easily get natural type of an expression,
when it is needed.  
- Finalized support for user-defined conversions from tuple literals.
- Made additional changes to the binding pipeline in order to have correct behavior around target typing.

CC @CyrusNajmabadi, @DustinCampbell For IDE changes and for the new API in SemanticModel.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionKind.cs:
```diff
@@ -12,7 +12,12 @@ internal enum ConversionKind : byte
-        ImplicitTuple,
+
+        /// <summary>
+        /// An implicit conversion from a tuple literal to a tuple or a tuple compatible type of matching cardinality.
```
Can you clarify what hte second case means? i.e. hte part about "a tuple compatible type of matching cardinality".  What's a code example where you get this?  Thanks!

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/CSharpExtensions.cs:
```diff
@@ -640,6 +640,22 @@ public static TypeInfo GetTypeInfo(this SemanticModel semanticModel, ExpressionS
+        public static ITypeSymbol GetNaturalType(this SemanticModel semanticModel, ExpressionSyntax expression, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var csmodel = semanticModel as CSharpSemanticModel;
+            if (csmodel != null)
```
Consider: use ?. operator here.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/Extensions.cs:
```diff
@@ -61,6 +61,18 @@ public static TypeInfo GetTypeInfo(this SemanticModel semanticModel, SyntaxNode
+        /// process of obtaining the semantic info.</param>
+        public static ITypeSymbol GetNaturalType(this SemanticModel semanticModel, SyntaxNode expression, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return semanticModel.GetNaturalType(expression, cancellationToken);
```
Isn't this just calling itself?  Or is this calling some internal instnace member that it has access to?

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/SemanticModel.cs:
```diff
@@ -104,6 +104,25 @@ internal SymbolInfo GetSymbolInfo(SyntaxNode node, CancellationToken cancellatio
+        /// <param name="expression">The syntax node to get the information for.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used to cancel the
+        /// process of obtaining the semantic info.</param>
+        internal ITypeSymbol GetNaturalType(SyntaxNode expression, CancellationToken cancellationToken = default(CancellationToken))
```
n/m.  i see now :)

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs:
```diff
@@ -3989,6 +3989,33 @@ class C
 {
     public void M()
     {
+        (((int, string))((1, ""hello""))).ToString();
```
Just to check.  But *ideally* we would not need this cast right?  @DustinCampbell to see what we'd need to do in the simplifier to make this work properly.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs:
```diff
@@ -3989,6 +3989,33 @@ class C
+{
+    public void M()
+    {
         ((1, ""hello"")).ToString();
```
Question.  The parens shouldn't be required here, right?   If that's the case then @DustinCampbell, our simplifier should unparenthesize tuples.

CyrusNajmabadi commented:
So i'm torn a little about the API addition to the semantic model.

Here's my general concern:

1. We have an existing entrypoint for giving us information about the types of an expression: GetTypeInfo.  It seems like the 'NaturalType' fits cleanly into that.  i.e. TypeInfo exposes Type, NaturalType, ConvertedType.
2. We did not go with GetType and GetConvertedType.  So having GetNaturalType seems inconsistent.

*However* i'm cognizant of the concern about how adding NaturalType to TypeInfo might make things more expensive (as the semantic model would now need to compute this all the time, even for consumers who don't need it).

I'd like @mattwar and @gafter  to weigh in on what we think the best approach is here.  Should we add nice simpel methods that mean our existing perf stays the same?  Or should we attempt to provide unified entrypoints, where we potentially make things more expensive.

I can see good arguments both ways.  And i'm not sure what our overall philosophy is here.

AlekseyTs commented:
@CyrusNajmabadi I guess, tuples is the first feature in C# that uses target typing, and has "natural" type in a true sense. There are existing features in VB though, lambdas, array literals, Nothing literal. They all have "natural" type, but it is not returned as TypeInfo.Type. The natural type only surfaces as ConvertedType in situations when there is no target type for the expression.

CyrusNajmabadi commented:
@AlekseyTs Totally understand :)

This comes down to simply a question of what we would prefer the ideal API to be.  I've added Neal/Matt to bring in their API hats to bear.  If this was internal only, this would be the right way to do things.  For public APIs though, i think sometimes we may end up making compromises (for example, around perf) to provide the 'right shaped' API.

I think we've crossed that path already.  We created SymbolInfo/TypeInfo to already combine values, even though that may be more expensive, and all those values may not be used by all consumers of the API.  So i think there's precedence to say that adding NaturalType into TypeInfo would be hte preferred approach.  But it's unclear to me what our principles are since this is an evolution of the API, and we haven't really run into this API evolution case yet.

AlekseyTs commented:
Adding NatiuralType to the TypeInfo sounds like a viable alternative to me. I was merely making an argument that the behavior of TypeInfo for tuples is consistent with its behavior for "similar" features in VB and, I think, we would want that consistency across languages.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
+        /// A literal that has a natural type should not add any errors to <paramref name="diagnostics"/>
+        /// during the process.
+        /// </summary>
+        private BoundTupleLiteral AdjustTupleLiteralBasedOnNaturalTypeInference(BoundTupleLiteral tupleLiteral, DiagnosticBag diagnostics, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```
It feels it would be much cheaper to compute natural type of a literal right when we bind it and already have types of elements available.
Deferring this recursive computation makes it much more complicated.

VSadov replied:
Having pre-computed natural type would also allow to shortcircuit most of the reclassifications.

gafter replied:
I do not think that NaturalType, if we end up with a spec that contains that concept, should be a separate API from TypeInfo.

AlekseyTs replied:
@gafter Could you clarify please. Would you prefer a new field in a TypeInfo structure, if this are the only two options available?

---
In reply to: [62959049](https://github.com/dotnet/roslyn/pull/11201#discussion_r62959049) [](ancestors = 62959049)

gafter replied:
@AlekseyTs My first preference is that we not introduce the concept into the compiler or the API. But if we must, it should be part of TypeInfo.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:
```diff
@@ -55,27 +55,28 @@ internal partial class Binder
                     return source;
                 }
             }
+            else if (ReclassifyExpression(ref source, syntax, conversion, isCast && !wasCompilerGenerated, destination, diagnostics))
```
In typical code in the vast majority of cases the natural type of the literal will match the type it is passed or assigned to. I do not think we need to reclassify in such cases

```C#
var x = (1,2); // this should not require recursive reclassification

int x = 1;
int y = 2;
foo((x, y));      // this should not require recursive reclassification

void foo<T>(T x)
{
}
```

AlekseyTs replied:
This is a different reclassification, the one that takes target type into account. Without doing this, nothing would work.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
+        /// Does tuple literal have natural type?
+        /// </summary>
+        private static bool TupleLiteralHasNaturalType(BoundTupleLiteral tupleLiteral)
+        {
```
This would also be much cheaper if tuple literal remembered its natural type right after binding.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
+        /// This could result in errors for certain expressions, when natural type cannot be inferred,
+        /// or in presence of other error conditions. 
+        /// 
+        /// An expression that has a natural type should not add any errors to <paramref name="diagnostics"/>
```
Sounds like this should be an assert

VSadov commented on src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -158,6 +158,8 @@ override Microsoft.CodeAnalysis.CSharp.Syntax.TupleTypeSyntax.Accept<TResult>(Mi
 override Microsoft.CodeAnalysis.CSharp.Syntax.WhenClauseSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
 static Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetDeclaredSymbol(this Microsoft.CodeAnalysis.SemanticModel semanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationPatternSyntax declarationSyntax, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.ISymbol
+static Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetNaturalType(this Microsoft.CodeAnalysis.SemanticModel semanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.ITypeSymbol
+static Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetSpeculativeNaturalType(this Microsoft.CodeAnalysis.SemanticModel semanticModel, int position, Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax expression, Microsoft.CodeAnalysis.SpeculativeBindingOption bindingOption) -> Microsoft.CodeAnalysis.ITypeSymbol
```
New public API or this is temporary?

VSadov commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -214,18 +214,18 @@ static void Main()
-  IL_001b:  ret
+  IL_0000:  ldc.i4.1
+  IL_0001:  ldc.i4.2
+  IL_0002:  newobj     ""System.ValueTuple<int, int>..ctor(int, int)""
```
This was in-place construction of a struct. Now it is construct-and-copy. Something went wrong with object creation nodes.

VSadov replied:
We should avoid copying except for rare cases where the difference is observable. (for example when we are in a try/catch etc... copying is required).
Codegen knows how and when to inline struct constructors. Possibly there is an extra conversion in the tree or something. - like for example instead of     ```local = new VT(..)``` it sees   ``` local = (cast) new VT()``` 

This applies to the other IL diffs below.

AlekseyTs replied:
Addressed in the 4th commit.

---
In reply to: [62953126](https://github.com/dotnet/roslyn/pull/11201#discussion_r62953126) [](ancestors = 62953126)

VSadov commented:
I am concerned about the introduced complexity here. 
It seems that what we mostly gain, is the ability to defer the computation of natural type. However in many cases we would need the natural type anyways or could benefit from knowing it.

Requiring all the assignments or argument passing or other kinds of tuple value transfer to unconditionally go through recursive reclassification looks a bit heavy. 

The reclassification is strictly needed only in cases involving nulls and lambdas or when natural type of the literal does not match the target type. Otherwise it is a semantical noop - you get a type that you could have from the beginning.

From the fact that the identity conversion is by far the most common conversion in existing code, I expect it to be the norm that natural types will nearly always match the destination types. We should optimize for that not for the opposite.

It feels it might be ok to use reclassify pattern, but only if we could bypass it when natural type matches the destination already.

gafter commented:
I am not happy about the logical complexity being added here, particularly as it relates to future maintenance, but I'm interested to review the language specification corresponding to these changes in order to help judge whether or not these changes should be integrated.

AlekseyTs commented:
@vsadov I would argue that in successful programs, we would rarely need to calculate natural type of a tuple, or reclassify it without a target type. So, the fact that we pre-compute will mostly be a waist of CPU. I prefer to not do premature optimization. We have a perf investigations on the list of things to do, if they will indicate that calculation of the "natural" type is a real issue, we can easily move to this approach, which, however, won't change anything in the way binding pipeline works and will be purely optimization.

AlekseyTs commented:
@gafter This change doesn't change behavior as it stand at the moment.

AlekseyTs commented:
I mean behavior of binding. Semantic model behavior obviously changing to make it consistent with other similar features.

AlekseyTs commented:
test prtest/win/dbg/unit32 please

AlekseyTs commented:
test prtest/win/dbg/unit64 please

AlekseyTs commented:
@dotnet/roslyn-infrastructure Is something wrong with Jenkins? test jobs are failing immediately after start.

VSadov commented:
@AlekseyTs  computing natural type at the time of literal binding is clearly cheaper than deferred recursive reclassification. This change trades one for another. It is a case of "premature generalization".

Reducing cases of reclassification only to the situations when it is necessary is important for the proposed approach. In a typical program we should not reclassify much at all. 
Any case where we reclassify to the natural type is a waste.

I agree that both old and new model can be made to behave exactly the same, but at this point it looks like the new model is both logically and computationally more complex. 
I am suggesting how it can be made closer computationally to the existing approach.

gafter commented:
I prefer the previous infrastructure for handling target typing.

AlekseyTs commented:
I don believe "computing natural type at the time of literal binding is clearly cheaper" is significant enough to have a noticeable impact. Visiting a tree is cheap, it is the same cost to build the type symbol. The only advantage could come from caching the result rather than recalculating it. However, calculating and caching information that is not used even once is a clear waist of CPU. For example, in case of ``` (byte, byte) x = (1, 2);```, the ```private BoundConversion ReclassifyTupleLiteral(BoundTupleLiteral tupleLiteral, DiagnosticBag diagnostics, ref HashSet<DiagnosticInfo> useSiteDiagnostics)``` is not called at all, we never calculate the "natural" type. The same is true for pretty much any other scenario when there is a target type. So, the benefit of unconditionally pre-computing the natural type is not that obvious to me. Given that, I am not going to implement this optimization just yet. 

---
In reply to: [218661496](https://github.com/dotnet/roslyn/pull/11201#issuecomment-218661496) [](ancestors = 218661496)

AlekseyTs commented:
>I prefer the previous infrastructure for handling target typing.

Are you talking specifically about SemanticModel changes? Is this what makes you uncomfortable?

---
In reply to: [218664340](https://github.com/dotnet/roslyn/pull/11201#issuecomment-218664340) [](ancestors = 218664340)

VSadov commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1142,43 +1142,23 @@
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="always"/>
+    <Field Name="Arguments" Type="ImmutableArray&lt;BoundExpression&gt;"/>
     <Field Name="ArgumentNamesOpt" Type="ImmutableArray&lt;string&gt;" Null="allow"/>
+    <Field Name="DeclaredOrTargetTupleTypeOpt" Type="TupleTypeSymbol" Null="allow"/>
```
This field needs a comment. 
It is not obvious what "DeclaredOrTargetTupleTypeOpt" represents

VSadov commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1142,43 +1142,23 @@
   The tuple literal is one which has not been converted to a target type.
   -->
-  <Node Name="BoundTupleLiteral" Base="BoundTupleExpression">
+  <Node Name="BoundTupleLiteral" Base="BoundExpression">
```
Previously we had target-typed and not target-typed literals represented by distinct nodes. The purpose was -
* make it impossible to target-type the same tuple literal twice
* trivially catch bugs if unconverted literals reach the emit phase
* reduce the size of bound nodes to minimally needed (converted tuple does not need to keep names, unconverted does not need two types)

That was similar to bound and unbound lambda nodes.

The new node combines it all together into one node.

Is it possible to keep converted and unconverted tuple nodes distinct?

AlekseyTs commented:
@VSadov That was my goal to get rid of the second node, because this what was the ideal from the very beginning. Two nodes was a compromise to reduce the cost of initial prototype.

- **Still true**: make it impossible to target-type the same tuple literal twice
- **Still true**: trivially catch bugs if unconverted literals reach the emit phase
- **Not a goal in my opinion**: reduce the size of bound nodes to minimally needed (converted tuple does not need to keep names, unconverted does not need two types)

Let's discuss this more offline.

AlekseyTs commented:
test vsi please

gafter commented:
> Are you talking specifically about SemanticModel changes? Is this what makes you uncomfortable?

I am referring both to the SemanticModel changes as well as the introduction into the C# compiler of the concept of reclassification as defined in the VB language specification. Whether we use that name or another name, I do not believe that concept should be introduced into the C# compiler.

---
In reply to: [218665679](https://github.com/dotnet/roslyn/pull/11201#issuecomment-218665679) [](ancestors = 218665679,218664340)

gafter commented:
:clock1:

jasonmalinowski commented:
@AlekseyTs There was an outage last night around the time you sent the message. Things should be good now. @dotnet-bot, restest vsi p2 please, this is a known failure that probably hadn't merged into future when you ran tests.

AlekseyTs commented:
This change became obsolete after recent LDM decisions. Everything that could easily be salvaged from this PR, has been committed in #11328.
